### ISSUE TITLE
HttpWebRequest doesn't send Digest authentication

### ISSUE THREAD

**Issue by filipnavara** [dotnet/corefx#32408]
Test program:
```c#
using System;
using System.Net;

namespace HttpTest
{
    class Program
    {
        static void Main(string[] args)
        {
            HttpWebRequest webRequest = WebRequest.CreateHttp("https://mail.emclient.com/webdav/test@emclient.com");
            webRequest.Method = "PROPFIND";
            webRequest.Credentials = new NetworkCredential("test@emclient.com", "<some password>");
            using (var response = webRequest.GetResponse())
            {
                // Should get here
            }
        }
    }
}
```

When targetting NetFX 4.5 the program successfully sends HTTP authentication headers and fulfills the request if the password is correct. This can easily be verified with Fiddler.

On .NET Core 2.1 and 3.0 (master from around last Friday) on Windows it doesn't even try to send the authentication headers.

**Comment by filipnavara** [comment]
Forcing WinHTTP handler using `AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);` makes it behave same as NetFX.

**Comment by filipnavara** [comment]
For reference, these are the HTTP headers sent by the server:
```
HTTP/1.1 401 Unauthorized
Server: IceWarp/12.1.1.3 RHEL7 x64
Date: Sat, 22 Sep 2018 15:22:09 GMT
Status: 401 Unauthorized
Set-Cookie: PHPSESSID=webdav8a07ed8692337c82525d2b79873511fee40fe9c3; path=/
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
Pragma: no-cache
MS-Author-Via: DAV
Access-Control-Allow-Origin: *
DAV: 1, access-control, calendar-access, calendar-auto-schedule, calendar-proxy, calendarserver-principal-property-search, extended-mkcol, addressbook
WWW-Authenticate: Digest realm="mail.emclient.com", qop="auth", algorithm=md5, nonce="fb97f659454cd25c8e38f913ca7b6d1a", opaque="4bc4f13929d28e0859ee285ad15429c6"
WWW-Authenticate: Basic realm="mail.emclient.com"
Content-Length: 0
Content-type: text/html; charset=UTF-8
```

**Comment by davidsh** [comment]
What version of .NET Core 2.1 are you using?  Please run "dotnet --info".

**Comment by filipnavara** [comment]
2.1.401, but it happens with builds from master too.

**Comment by filipnavara** [comment]
I have a suspiction that it's these lines:

https://github.com/dotnet/corefx/blob/4f7690ad97ea1772433fcfeafc4b0c26f0fb80a2/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.Digest.cs#L52-L56

The algorithm comparison is done in case-sensitive manner and the casing in CoreFX is `MD5`, while in the response it is `md5`.

**Comment by davidsh** [comment]
That could be a bug.  The comparison should be done in a 'case insensitive' manner, IMHO.  But the RFC doesn't state it explicitly.

See:
https://tools.ietf.org/html/rfc2617#section-3.2.1

It's also possible that the RFC implied it should be uppercase "MD5" but some servers aren't fully compliant with that.  It would be useful to see what other browsers etc. do regarding this.  And also do a deep dive on the RFC to see if 'case insensitive' comparisons are implied.

cc: @geoffkizer @stephentoub @mconnew

**Comment by davidsh** [comment]
cc: @karelz

**Comment by geoffkizer** [comment]
Yeah, the RFC doesn't seem to be clear here. 

Given that WinHttp and NETFX both do a case insensitive compare, it seems reasonable to match their behavior here.

**Comment by filipnavara** [comment]
It's not just WinHTTP and NetFX. This server was used for many years and it was tested against all major browsers on many platforms, including macOS (Safari, Chrome, Apple Calendar, Apple Contacts) and iPhone.

I do agree that the wording in the RFC doesn't state explicitly that it is case-insensitive, but the status quo is that the implementations treat it as such and it is implied through the ABNF syntax. ABNF explicitly states "ABNF strings are case insensitive and the character set for these strings is US-ASCII." The case-sensitive strings in ABNF were introduced as an extension many years after the original specification. RFC 2831, which maps Digest authentication into the SASL framework even explicitly uses lower case "md5-sess" value.

**Comment by filipnavara** [comment]
Yes, but I am quite busy at the moment. Maybe tomorrow or early next week.

**Comment by davidsh** [comment]
Fixed in master with PR #32632

**Comment by davidsh** [comment]
@karelz Should we consider this for release/2.1.x ?

**Comment by karelz** [comment]
Impact: Some servers don't work with .NET Core 2.1+ client (Digest authentication). Affects both HttpWebRequest and HttpClient APIs.
We should fix it in 2.2.

**Comment by davidsh** [comment]
Fixed in release/2.2 with PR #32920
